modules/core/04-channel/keeper/upgrade.go

```diff
@@ -85,24 +82,245 @@ func (k Keeper) WriteUpgradeInitChannel(
+			// set to the max of the two
+			if counterpartyUpgradeSequence > upgradeSequence {
+				upgradeSequence = counterpartyUpgradeSequence
+				k.SetUpgradeSequence(ctx, portID, channelID, upgradeSequence)
+			}
+
+			k.SetUpgradeErrorReceipt(ctx, portID, channelID, errorReceipt)
+			return 0, "", errorsmod.Wrapf(types.ErrUpgradeAborted, "upgrade aborted, error receipt written for upgrade sequence: %d", errorReceipt.Sequence)
+		}
+
+		// if there is a crossing hello, i.e an UpgradeInit has been called on both channelEnds,
+		// then we must ensure that the proposedUpgrade by the counterparty is the same as the currentChannel
+		// except for the channel state (upgrade channel will be in TRYUPGRADE and current channel will be in INITUPGRADE)
+		// if the proposed upgrades on either side are incompatible, then we will restore the channel and cancel the upgrade.
+		channel.State = types.TRYUPGRADE
+
+		if !reflect.DeepEqual(channel, proposedUpgradeChannel) {
+			// TODO: log and emit events
+			if err := k.RestoreChannelAndWriteErrorReceipt(ctx, portID, channelID, upgradeSequence, types.ErrInvalidChannel); err != nil {
+				return 0, "", errorsmod.Wrap(types.ErrUpgradeAborted, err.Error())
+			}
+			return 0, "", errorsmod.Wrap(types.ErrUpgradeAborted, "proposed upgrade channel did not equal expected channel")
+		}
+
+		// retrieve restore channel to return previous version
+		restoreChannel, found := k.GetUpgradeRestoreChannel(ctx, portID, channelID)
+		if !found {
+			// (this case should be unreachable): write error receipt for upgrade sequence and abort / cancel upgrade
+			errorReceipt := types.NewErrorReceipt(upgradeSequence, err)
+			k.SetUpgradeErrorReceipt(ctx, portID, channelID, errorReceipt)
+			return 0, "", errorsmod.Wrapf(types.ErrChannelNotFound, "upgrade aborted, error receipt written for upgrade sequence: %d", upgradeSequence)
+		}
+
+		return upgradeSequence, restoreChannel.Version, nil
+	default:
+		return 0, "", errorsmod.Wrapf(types.ErrInvalidChannelState, "expected one of [%s, %s] but got %s", types.OPEN, types.INITUPGRADE, channel.State)
+	}
+
+	return upgradeSequence, channel.Version, nil
+}
+
+// WriteUpgradeTryChannel writes a channel which has successfully passed the UpgradeTry handshake step.
+// An event is emitted for the handshake step.
+func (k Keeper) WriteUpgradeTryChannel(
+	ctx sdk.Context,
+	portID,
+	channelID,
+	proposedUpgradeVersion string,
+	upgradeSequence uint64,
+	channelUpgrade types.Channel,
+) {
+	defer telemetry.IncrCounter(1, "ibc", "channel", "upgrade-try")
+
+	channel, found := k.GetChannel(ctx, portID, channelID)
+	if !found {
+		panic(fmt.Sprintf("failed to retrieve channel %s on port %s", channelID, portID))
+	}
+
+	// assign directly the fields that are modifiable.
+	// counterparty fields may not be changed.
+	channel.State = types.TRYUPGRADE
+	channel.Version = proposedUpgradeVersion
+	channel.Ordering = channelUpgrade.Ordering
+	channel.ConnectionHops = channelUpgrade.ConnectionHops
```

Comment:
Nit: this code updating the channel fields [is duplicated in `WriteUpgradeInitChannel`](https://github.com/cosmos/ibc-go/blob/cian/issue%231619-implement-channelupgradetry-functionality-for-04-channel/modules/core/04-channel/keeper/upgrade.go#L94-L99). Maybe we can have a method on the channel end struct to update all upgradable fields?

Reply:
Are you happy if we introduce a helper function in a follow up?

Reply:
Yes, more than happy! :)